golang/pkg/crane/k8s/service_monitor.go

```diff
@@ -0,0 +1,118 @@
+package k8s
+
+import (
+	"context"
+
+	"github.com/rs/zerolog/log"
+
+	v1 "github.com/dyrector-io/dyrectorio/golang/api/v1"
+	"github.com/dyrector-io/dyrectorio/golang/pkg/crane/config"
+
+	monitoringapiv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/client/versioned/typed/monitoring/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+)
+
+type ServiceMonitor struct {
+	Ctx       context.Context
+	Client    *monitoringv1.MonitoringV1Client
+	appConfig *config.Configuration
+}
+
+const (
+	ServiceMonitorKind    = "ServiceMonitor"
+	ServiceMonitorVersion = "monitoring.coreos.com/v1"
+)
+
+// service monitor is spawned per deployment
+func NewServiceMonitor(ctx context.Context, cli *Client) *ServiceMonitor {
+	if !cli.VerifyAPIResourceExists(ServiceMonitorVersion, ServiceMonitorKind) {
+		return nil
+	}
+
+	restConf, err := cli.GetRestConf()
+	if err != nil {
+		log.Panic().Err(err).Stack().Send()
+	}
+
+	// todo(nandor-magyar): fix error handling here, and with every sub-component
+	smClient, _ := NewServiceMonitorClient(restConf)
+
+	return &ServiceMonitor{
+		Ctx:       ctx,
+		Client:    smClient,
+		appConfig: cli.appConfig,
+	}
+}
+
+// Deploy firstPort is used as a fallback if only path is provided, the first port is used
+func (sm *ServiceMonitor) Deploy(namespace, serviceName string, metricParams v1.Metrics, firstPort string) error {
+	newMonitor := createServiceMonitorSpec(serviceName, metricParams, firstPort)
+	clientSet := sm.Client.ServiceMonitors(namespace)
+
+	oldServiceMonitor, err := clientSet.Get(sm.Ctx, serviceName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		_, err = clientSet.Create(sm.Ctx, newMonitor, metav1.CreateOptions{
+			FieldManager: sm.appConfig.FieldManagerName,
+		})
+		if err != nil {
+			return err
+		}
+	} else if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	} else {
+		newMonitor.ResourceVersion = oldServiceMonitor.ResourceVersion
+		_, err := clientSet.Update(sm.Ctx, newMonitor, metav1.UpdateOptions{
+			FieldManager: sm.appConfig.FieldManagerName,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// cleanup ignores errors, expected to run if no metrics is were defined
+func (sm *ServiceMonitor) Cleanup(namespace, serviceName string) {
+	clientSet := sm.Client.ServiceMonitors(namespace)
+	_ = clientSet.Delete(sm.Ctx, serviceName, metav1.DeleteOptions{})
+}
+
+func createServiceMonitorSpec(name string,
+	metricParams v1.Metrics,
+	defaultPortName string,
+) *monitoringapiv1.ServiceMonitor {
+	portName := metricParams.Port
+
+	if portName == "" {
+		portName = defaultPortName
+	}
+
+	metricsPath := metricParams.Path
+	if metricsPath == "" {
+		metricsPath = "/metrics"
```

Comment:
I think extracting this to a const would be nice (`defaultMetricsPath` or something).

Reply:
Singular use, unnecessary obstruction.